Umb8 - dashboard created for editing robots txt.

emmaburstow opened the pull request:
This work submitted for your perusal contains the following things:

We have built the dashboard for the package,
put in a textarea,
gotten the contents of the robots.txt coming back into the text area, 
gotten it to write anything we add in to the robots.txt file, 
got notifications working to show if it's successful, 
put in a button that reloads the filecontents into the textarea. 

Still to do:
We like the idea that the textarea could be one of those ones that have line numbers in. 

Enjoy!

LottePitcher committed:
add starting solution, classes project with post build events to copy package files

LottePitcher committed:
Editorial changes to GETTINGSTARTED.md

LottePitcher committed:
Another tweak to GETTINGSTARTED.md

Commit:
add package.manifest file and attempt to add folder copy functionality working

LottePitcher committed:
Remove rogue $ from PostBuildEvents.bat

@MMasey just a rogue $ in there :( then should work.  I've updated my PostBuildEvents.bat accordingly and yup, package.manifest now in my test website! Hopefully we won't need to change the .bat file any more after this!

Commit:
add initial dashboard to settings section. Not sure if it is in the correct place, just trying to get make it visible.

Commit:
update GETTINGSTARTED.md to include info about changing the ClientDependency version in order to get JS files to work

LottePitcher committed:
add dictionary item so dashboard link text appears properly

LottePitcher committed:
I see other package.manifests do include 'name' (as does documentation) even tho not in official json schema, so put this back in, sorry!

LottePitcher committed:
Merge branch 'startumb8LP' into umb8

Commit:
Angular controller passing string to the dashboard, button and text area in.

Commit:
Api created to get text

Commit:
Resource added for api get method

Commit:
Returning data from text file to view, some styling tidied up

Commit:
Model created so we can do more in the apicontroller

LottePitcher committed:
API data being added to view model, not over-writing (and breaking!) it

LottePitcher committed:
Update GETTINGSTARTED.md with reminder re debug to true to disable client dependency caching (as I did not know that!)

LottePitcher committed:
organise js controller in similar way to models builder, including a 'loading' flag

LottePitcher committed:
update view layout/styling to be more consistent with core dashboards

LottePitcher committed:
oops: forgot to update save function to post the 'data' property of view model to API controller

LottePitcher committed:
reload button if file exists, and button says Create if missing, or Update if existing

LottePitcher committed:
removing redundant div from view

Commit:
RobotsTxt filecontents updating when submitted

Commit:
Notifications service used to give success message

leekelleher commented:
:tada::tada::tada::tada::tada::tada::tada::tada::tada::tada::tada::tada::tada::tada::tada::tada::tada::tada::tada::tada:

Feels like Christmas! 🎁 

I'll make time this week to review it and feedback. 😄

LottePitcher commented:
Thanks @leekelleher (and a merry christmas!!!).  Any comments appreciated, be it relating to solution structure, C# or angular! 

Perhaps you also have an opinion on whether it really should be a tree node, instead of a new dashboard item. There's a card on Trello discussing that.

There is a new GETTINGSTARTED.md file that explains how you can get the project post build copying to your own website. Although it shouldn't fall over if you don't follow those instructions, it just won't do anything on post build, and you can do your usual own XCOPY malarkey!

LottePitcher committed:
remove dummy.cs as now have actual C# in the dll!

LottePitcher committed:
exclude nuget packages from git

Commit:
fix: update project file to fix build error on fresh install

Commit:
feat: add initial default robots.txt value when not file exists.

Commit:
wip: copied over some code from core that enables the ace editor. CURRENTLY BROKEN

leekelleher commented:
OK, I've (finally) got around to doing an initial review. I have feedback - all good, don't worry! 😃
How to best feed it back to you all? I could do it as a PR to this PR? (I know, very meta)
or go through line-by-line on this PR? But I fear that I may appear slightly obnoxious. 😝

leekelleher commented:
re: the XCOPY, turns out that in copying over v8.0.0 DLLs to a v8.0.1 site, well Umbraco doesn't like that at all and it borks the database! 😱  I adjusted the "Copy Local" in VS so that the build doesn't copy all the Umbraco DLLs to the "/bin/Debug" folder.  _Just one of the joys of package dev!_ 😆

leekelleher commented on CONTRIBUTING.md:
````diff
@@ -117,6 +117,8 @@ project:
    git checkout -b <topic-branch-name>
    ```
 
+   If you need guidance on how to work with this repository, please refer to [Getting started](GETTINGSTARTED.md)
````
✏️ Missing full-stop at end of sentence.

leekelleher commented on src/Our.Umbraco.RobotsTxtEditor/Controllers/RobotsTxtEditorApiController.cs:
```diff
@@ -0,0 +1,59 @@
+using Umbraco.Web.WebApi;
+using Umbraco.Web.WebApi.Filters;
+using Our.Umbraco.RobotsTxtEditor.Models;
+using System;
```
🛁 Namespace sorting. In VS use "Remove and Sort Usings" option (`CTRL+R`, `CTRL+G`)

leekelleher commented on src/Our.Umbraco.RobotsTxtEditor/Controllers/RobotsTxtEditorApiController.cs:
```diff
@@ -0,0 +1,59 @@
+        [HttpGet]
+        public RobotsTxtEditorModel GetRobotsText()
+        {
+            var filePath = HttpContext.Current.Server.MapPath("~/robots.txt");
```
We can use Umbraco's `IOHelper` here instead of `HttpContext.Current.Server`. Less reliance on `HttpContext`, and lets Umbraco deal any gotchas.

leekelleher commented on src/Our.Umbraco.RobotsTxtEditor/Controllers/RobotsTxtEditorApiController.cs:
```diff
@@ -0,0 +1,59 @@
+        {
+            var filePath = HttpContext.Current.Server.MapPath("~/robots.txt");
+
+            if (!File.Exists(filePath))
```
:neckbeard: Opinion bit here. In recent years I've preferred `foo == false` over `!foo`, improved readability (for me).
Nothing bad about using !negation, the compiler doesn't care, produces the same binary code 😃

leekelleher commented on src/Our.Umbraco.RobotsTxtEditor/Controllers/RobotsTxtEditorApiController.cs:
```diff
@@ -0,0 +1,59 @@
+            var filePath = HttpContext.Current.Server.MapPath("~/robots.txt");
+
+            //create the file and then write to it
+            using (StreamWriter sw = File.CreateText(filePath))
```
:neckbeard: Consistency. On L31 `var` is used, here it's explicit.  Personally, I've been going with `var`, but appreciate some folk prefer the old way.

leekelleher commented on src/Our.Umbraco.RobotsTxtEditor/Controllers/RobotsTxtEditorApiController.cs:
```diff
@@ -0,0 +1,59 @@
+                    FileExists = true,
+                    FileContents = reader.ReadToEnd()
+                };
+                return vm;
```
If we're not doing anything else with `vm` here, we could return the model immediately. Saves an allocation.

```csharp
return new RobotsTxtEditorModel { ... };
```

leekelleher commented on src/Our.Umbraco.RobotsTxtEditor/Controllers/RobotsTxtEditorApiController.cs:
```diff
@@ -0,0 +1,59 @@
+                sw.WriteLine(vm.FileContents);
+            }
+           
+            return true;
```
🤔 Wondering if we should consider any failures/exceptions here? Would we ever return false? _and what would the frontend do?_

leekelleher commented on src/Our.Umbraco.RobotsTxtEditor/Models/RobotsTxtEditorModel.cs:
```diff
@@ -0,0 +1,15 @@
+    {
+        public string FileContents { get; set; }
+        public bool FileExists { get; set; }
+        
```
🛁 Whitespace formatting. Apologies for my OCD 😉 

...and remove unused namespaces :neckbeard:

leekelleher commented on src/Our.Umbraco.RobotsTxtEditor/Properties/AssemblyInfo.cs:
```diff
@@ -0,0 +1,36 @@
+// by using the '*' as shown below:
+// [assembly: AssemblyVersion("1.0.*")]
+[assembly: AssemblyVersion("1.0.0.0")]
+[assembly: AssemblyFileVersion("1.0.0.0")]
```
Most folk ignore this file, but I obsess about it... even down to encoding the copyright symbol &copy; - there's an odd story behind this. If this file is saved as ASCII, (not UTF8) then the &copy; appears as a square glyph in the info-box when you view Properties on the DLL itself, (in Windows Explorer) ... _odd story, but it bothered me._ 🤣 

For OCD reference, here's our AssemblyInfo for Stacked Content: https://github.com/umco/umbraco-stacked-content/blob/develop/src/Our.Umbraco.StackedContent/Properties/AssemblyInfo.cs#L9

leekelleher commented on src/Our.Umbraco.RobotsTxtEditor/Web/UI/App_Plugins/RobotsTxtEditor/js/robotstxteditorController.js:
```diff
@@ -0,0 +1,53 @@
+                vm.data = response.data;
+
+                if(vm.data.FileExists === false) {
+                    // TODO: See if there is a nice way of doing this whilst maintaining correct format
```
🤔 Wondering if this could be handled in the server-side controller? e.g. if the file doesn't exist, make one. 

In the old version of this package, I used an [embedded resource to store the `robots.txt`](https://github.com/leekelleher/umbraco-robotstxt-editor/blob/master/Our.Umbraco.Tree.RobotsTxt/Our.Umbraco.Tree.RobotsTxt.csproj#L72), and pull it out from the assembly when we needed to create a new file: https://github.com/leekelleher/umbraco-robotstxt-editor/blob/master/Our.Umbraco.Tree.RobotsTxt/umbraco/plugins/robots-txt/editRobotsTxtFile.aspx.cs#L139-L146

leekelleher commented on src/Our.Umbraco.RobotsTxtEditor/Controllers/RobotsTxtEditorApiController.cs:
```diff
@@ -0,0 +1,59 @@
+
+        [HttpPost]
+        public bool SaveRobotsText(RobotsTxtEditorModel vm)
+        {
```
🤔 Validation?

Here's how the old version did it: https://github.com/leekelleher/umbraco-robotstxt-editor/blob/master/Our.Umbraco.Tree.RobotsTxt/umbraco/plugins/robots-txt/editRobotsTxtFile.aspx.cs#L49-L67

leekelleher commented on src/Our.Umbraco.RobotsTxtEditor/Web/UI/App_Plugins/RobotsTxtEditor/js/robotstxteditorController.js:
```diff
@@ -0,0 +1,53 @@
+                console.log("API post returned " + success);
+                if (success) {
+                    vm.data.FileExists = true;
+                    notificationsService.success('Saved', 'Text saved to Robots.txt');
```
:neckbeard: Consistency of quotes. This is me being picky, sorry. Either use single or double throughout the JS file.

leekelleher commented on src/Our.Umbraco.RobotsTxtEditor/Web/UI/App_Plugins/RobotsTxtEditor/views/robotstxteditor.html:
```diff
@@ -0,0 +1,40 @@
+                        <!-- TODO: replace with 'posh' textarea with line numbers (like editing styles)-->
+                        <textarea ng-model="vm.data.FileContents"></textarea>
+                    </p>
+                    <p>
```
🤔 Wondering if we could add something _fun_ here? Maybe a robot cartoon saying "create me?"
Oooo!, maybe we could commission a Doodlebeard? // @mattbrailsford 😍 (no pressure Matt, just spit-balling)

emmaburstow replied:
Hey @mattbrailsford! Happy Easter! How's that Robot doodle coming along? 😉

leekelleher commented on src/Our.Umbraco.RobotsTxtEditor/Web/UI/App_Plugins/RobotsTxtEditor/views/robotstxteditor.html:
```diff
@@ -0,0 +1,40 @@
+                </div>
+                <div class="umb-robotstxteditor-notes">
+                    <!-- TODO: add button to insert 'standard' rules for an Umbraco website -->
+                    <!-- TODO: add other buttons / helpful functionality -->
```
📓 Other features from the old version were...

1. Insert User-Agent rule
2. Insert Disallow rule
3. Comment out rules
4. Uncomment rules

**JS**: https://github.com/leekelleher/umbraco-robotstxt-editor/blob/master/Our.Umbraco.Tree.RobotsTxt/umbraco/plugins/robots-txt/editRobotsTxtFile.aspx#L5-L32
**C#**: https://github.com/leekelleher/umbraco-robotstxt-editor/blob/master/Our.Umbraco.Tree.RobotsTxt/umbraco/plugins/robots-txt/editRobotsTxtFile.aspx.cs#L87-L109

leekelleher commented on src/Our.Umbraco.RobotsTxtEditor/Web/UI/App_Plugins/RobotsTxtEditor/views/robotstxteditor.html:
```diff
@@ -0,0 +1,40 @@
+                    <div ng-show="vm.data.FileExists">
+                        <umb-button type="button"
+                                    label="Reload from disk"
+                                    action="vm.reloadData()">
```
🤔 Wondering if we need a Reload button. Are we expecting people to be editing it in both places at the same time? _and hoping that the person in the back-office will hit the reload button?_

LottePitcher replied:
@leekelleher Ok so yes I can see the confusion here. I  was thinking of it more as a 'cancel changes' option.  i.e. i've done some changes and got confused, so i want to reload it from disk and start again.  So perhaps it's a 'cancel' button alongside the Save?  Maybe 'Cancel' buttons aren't very 'umbraco', and they assume a user would click away and back again to effect a cancel.  I'm don't feel strongly either way.  It was just nice to get a button working ;-)

leekelleher replied:
Ah ok, gotcha... "Revert to saved" (for want of a better phrase)

emmaburstow replied:
I like Revert to saved too so have changed this :)

leekelleher commented on src/PostBuildEvents.bat:
```diff
@@ -0,0 +1,17 @@
+xcopy /s /y "%2%3.pdb" "{YOUR-WEBSITE-LOCATION}\bin\"
+
+REM Copy package front-end files
+xcopy /s /y "%4Web\UI\*.*" "{YOUR-WEBSITE-LOCATION}"
```
I'm loving this file, thank you! 🙏 

🤔 Wondering if we could use a local variable here for the `{YOUR-WEBSITE-LOCATION}`?
e.g. `SET /P WEBSITE_PATH=C:\path\to\my\umbraco8\website`

_I'll have a play around._ 😃

LottePitcher replied:
@leekelleher would love to see if this could be a bit easier!

leekelleher replied:
Just added this in.

leekelleher reviewed (changes requested):
I went for a big old review on here.

Hopefully it should go without saying, none of my commends/feedback is personal, most of it comes down to my personal coding preferences. Take them with a pinch of salt.

Overall, I'm very happy.

- The dashboard approach is a good choice - better suited for v8 style.
- I think we can do something fun with the layout, I like the idea of a robot character, _(but then just had a shudder of Clippy!)_
- Not sure if we need a reload button. _Happy to discuss that._

I really do appreciate the time, effort and energy you've all put into this! ❤️

leekelleher committed:
Removed .gitkeep file

It was only there as a placeholder for the folder.

leekelleher committed:
Remove and Sort Usings namespaces

leekelleher committed:
Whitespace formatting / housekeeping

leekelleher committed:
Sets the "Copy Local" to false

In Visual Studio, by default, when you build a project, all the assembly
references are copied over to the target directory.

This is problematic for Umbraco packages, as if you XCOPY all the
Umbraco assemblies over to an existing Umbraco installation, then you
may end up with a version mismatch.
e.g. v8.0.0 assemblies in a v8.0.1 installation.
and Umbraco does not like that at all.

So we need to make sure that the Umbraco assemblies aren't in the
target directory.

leekelleher committed:
AssemblyInfo clean up

leekelleher committed:
Switched to use `IOHelper`.

Using `HttpContext` directly, may have complications
- that Umbraco can already handle.

leekelleher committed:
Coding style to switch from negation to explicit

leekelleher committed:
Returns the view-model directly

If we aren't manipulating the value further, then we can return
it straightaway.

leekelleher committed:
Coding consistency, using `var` keyword

leekelleher committed:
JS quoting consistency, using double-quotes

MMasey committed:
Merge pull request #6 from leekelleher/wip-umb8-patch1

[Umb8] Lee's code review patch

Commit:
feat: add correct content to editor on load

Commit:
Merge remote-tracking branch 'origin/umb8' into umb8-fancy-editor

# Conflicts:
#	src/Our.Umbraco.RobotsTxtEditor/Web/UI/App_Plugins/RobotsTxtEditor/js/robotstxteditorController.js
#	src/Our.Umbraco.RobotsTxtEditor/Web/UI/App_Plugins/RobotsTxtEditor/views/robotstxteditor.html

Commit:
fix: unfinished merge and consitency of quotation marks

Commit:
Change of reload button text

Commit:
feat: reconfigure layout of buttons for robots.txt editor

Commit:
improvement: remove unused aceEditor keybindings

Commit:
feat: update editor to include heading text

Commit:
ricidulously messy words for guidance to be tidied up and put partially into comments in the robots file

Commit:
feat: add helper text to created robots.txt file and a remove it from dashboard.

Commit:
feat: light grey border added to the fancy editor

Commit:
Merge branch 'umb8-fancy-editor' into umb8

# Conflicts:
#	src/Our.Umbraco.RobotsTxtEditor/Web/UI/App_Plugins/RobotsTxtEditor/views/robotstxteditor.html

Commit:
feat: update save file response with validation and response model

Commit:
feat: update default robots.txt help text and format

Commit:
fix: update robots save method so that it only updates if the contents are valid

Commit:
feat: add validation error messages

Commit:
fix: bug where saving the file would keep adding a new line

LottePitcher committed:
Merge pull request #7 from leekelleher/feature/robots-validation

Feature/robots validation

LottePitcher committed:
add editor.config file (taken from Umbraco source)

leekelleher committed:
Merge pull request #9 from leekelleher/editorconfig

Addition of Editorconfig

LottePitcher committed:
moving .editorconfig file to the root of the solution

Commit:
Changes to button placement and alert used to notify user they do not have robots file yet

LottePitcher committed:
remove now redundant div tags, class declarations, and move note about default content to the alert that shows if file does not exist

LottePitcher committed:
remove now redundant css from stylesheet

LottePitcher committed:
remove commented out code from robotstxteditorController.js and simplify the specification of default contents

LottePitcher committed:
remove console.log message

leekelleher committed:
Merge pull request #10 from leekelleher/umb8-button-and-alert

UX changes and remove redundant css/js/html

leekelleher committed:
Grammar - added full stop.

leekelleher committed:
Updated PostBuildEvents.bat

Added the `WEBSITE_PATH` variable.

LottePitcher committed:
upgrade Umbraco to v8.1.0

LottePitcher committed:
removing the umb-editor-container div to remove the empty space at bottom of view

LottePitcher committed:
remove flagging when 'dirty' as not being reset properly, and stop editor loading all highlighted

leekelleher committed:
Set assembly references "Copy Local" to False

leekelleher committed:
Added build script for creating the packages

Copied over from https://github.com/umco/umbraco-property-list

leekelleher committed:
Updated README

Had one of those "what is 'Our'?" moments.

LottePitcher committed:
customising nuget and umbraco package settings for robots.txt

LottePitcher committed:
update build scripts to work with VS 2019 (any edition)

emmaburstow committed:
Merge pull request #11 from leekelleher/upgrade-to-v81

Upgrade to v81

leekelleher committed:
Upgraded Umbraco dependency to v8.2.0

LottePitcher committed:
Merge pull request #14 from leekelleher/upgrade-to-v82

Upgraded Umbraco dependency to v8.2.0

leekelleher committed:
Merge pull request #12 from leekelleher/ux-improvements

UX improvements

leekelleher committed:
Merge branch 'umb8' into umb8-build-scripts

leekelleher committed:
Updated build script reference to Umbraco v8.2 dependency

leekelleher committed:
New logo for v8 version

Happy Robot!

emmaburstow committed:
Merge pull request #15 from leekelleher/umb8-logo

New logo for v8 version

LottePitcher committed:
Merge pull request #13 from leekelleher/umb8-build-scripts

Added build script for creating the packages

leekelleher committed:
Added smaller logo

Updated logo URL for the package(s).
